jaxlib/lapack.pyx

```diff
@@ -383,3 +384,125 @@ def jax_potrf(c, a, lower=False):
           Shape.array_shape(np.int32, (), ()),
           Shape.array_shape(dtype, (n, n), (0, 1)),
       ))
+
+
+
+# ?gesdd: SVD decomposition
+
+cdef void lapack_sgesdd(void* out_tuple, void** data) nogil:
+  cdef int32_t job_opt = (<int32_t*>(data[0]))[0]
+  cdef int m = (<int32_t*>(data[1]))[0]
+  cdef int n = (<int32_t*>(data[2]))[0]
+  cdef float* a_in = <float*>(data[3])
+
+  cdef void** out = <void**>(out_tuple)
+  cdef float* a_out = <float*>(out[0])
+  cdef float* s = <float*>(out[1])
+  cdef float* u = <float*>(out[2])
+  cdef float* vt = <float*>(out[3])
+  cdef int* info = <int*>(out[4])
+
+  if a_out != a_in:
+    memcpy(a_out, a_in, m * n * sizeof(float))
+
+  # define appropriate job code
+  cdef char jobz = 'A'
+  if job_opt_compute_uv == 0:
+    jobz = 'N'
+  else:
+    if job_opt_some == 1:
+      jobz = 'S'
+
+  cdef int lda = m
+  cdef int ldu = m
+  cdef int ldvt = n
+
+  cdef int* iwork = <int *> malloc(min(m, n) * sizeof(int))
```

Comment:
You need to free the malloc-ed workspace buffers (same everywhere).

There is another alternative, which is to have XLA allocate the workspace buffers, which you can do by making them dummy outputs of the XLA custom call. However, then you need to do something else to figure out the workspace sizes, e.g., use the formula in the LAPACK docs, or add a separate method that performs the query.

Having XLA allocate the buffers would be slightly preferable, since then XLA can allocate all temporary buffers for the entire computation as a slab, saving malloc/free overhead. But it doesn't matter that much for a first pass, so up to you if you want to do this.